numba/jitclass/base.py

```diff
@@ -0,0 +1,362 @@
+    for k, v in methods.items():
+        jitmethods[k] = njit(v)
+
+    class_type = class_ctor(cls, spec, jitmethods)
+    cls = JitClassType(cls.__name__, (cls,), dict(class_type=class_type,
+                                                  __doc__=docstring))
+
+    # Register resolution of the class object
+    typer = CPUTarget.typing_context
+    typer.insert_global(cls, class_type)
+
+    # Register class
+    backend = CPUTarget.target_context
+    builder(class_type, methods, typer, backend).register()
+
+    return cls
+
+
+def _drop_ignored_attrs(dct):
+    # ignore anything defined by object
+    drop = set(['__weakref__',
+                '__module__',
+                '__dict__'])
+    for k, v in dct.items():
+        if isinstance(v, (pytypes.BuiltinFunctionType,
+                          pytypes.BuiltinMethodType)):
+            drop.add(k)
+        elif getattr(v, '__objclass__', None) is object:
+            drop.add(k)
+
+    for k in drop:
+        del dct[k]
+
+
+class ClassBuilder(object):
+    """
+    A jitclass builder for mutable jitclasses.
+    """
+    instance_type_class = types.ClassInstanceType
+    per_backend_registry = {}
+
+    def __init__(self, class_type, methods, typer, backend):
+        self.class_type = class_type
+        self.methods = methods
+        self.typer = typer
+        self.backend = backend
+        self.register_once_per_class()
+
+    def register_once_per_class(self):
+        registry = imputils.Registry()
+        if (type(self), self.backend) not in self.per_backend_registry:
+            self.implement_constructor(registry)
+            self.implement_attribute(registry)
+            self.backend.insert_func_defn(registry.functions)
+            self.backend.insert_attr_defn(registry.attributes)
+            self.per_backend_registry[type(self), self.backend] = registry
+
+    def register(self):
+        """
+        Register to the frontend and backend.
+        """
+        # outer = self
+
+        class_type = self.class_type
+        instance_type = class_type.instance_type
+
+        class ConstructorTemplate(templates.AbstractTemplate):
+            key = class_type
+
+            def generic(self, args, kws):
+                ctor = instance_type.jitmethods['__init__']
+                boundargs = (instance_type.get_reference_type(),) + args
+                template, args, kws = ctor.get_call_template(boundargs, kws)
+                sig = template(self.context).apply(args, kws)
+                out = templates.signature(instance_type, *sig.args[1:])
+                return out
+
+        self.typer.insert_function(ConstructorTemplate(self.typer))
+        self.implement_frontend(instance_type)
+        self.implement_backend(instance_type)
+
+    def implement_frontend(self, instance_type):
+        self.implement_attribute_typing(instance_type)
+
+    def implement_attribute_typing(self, instance_type):
+        class ClassAttribute(templates.AttributeTemplate):
```

Comment:
I still wonder why you can't define this at the top-level (with `key = types.ClassInstanceTyoe`).